src/deluge/io/midi/midi_follow.cpp

```diff
@@ -0,0 +1,700 @@
+		if (patchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+			paramKind = Param::Kind::PATCHED;
+			paramID = patchedParamShortcuts[xDisplay][yDisplay];
+		}
+		else if (unpatchedParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+			paramKind = Param::Kind::UNPATCHED_SOUND;
+			paramID = unpatchedParamShortcuts[xDisplay][yDisplay];
+		}
+		else if (globalEffectableParamShortcuts[xDisplay][yDisplay] != kNoParamID) {
+			paramKind = Param::Kind::UNPATCHED_GLOBAL;
+			paramID = globalEffectableParamShortcuts[xDisplay][yDisplay];
+		}
+
+		if (display->haveOLED()) {
+			DEF_STACK_STRING_BUF(popupMsg, 40);
+
+			const char* name = getParamDisplayName(paramKind, paramID);
+			if (name != l10n::get(l10n::String::STRING_FOR_NONE)) {
+				popupMsg.append("Can't control: \n");
+				popupMsg.append(name);
+			}
+			display->displayPopup(popupMsg.c_str());
+		}
+		else {
+			display->displayPopup(l10n::get(l10n::String::STRING_FOR_PARAMETER_NOT_APPLICABLE));
+		}
+	}
+
+	return modelStackWithParam;
+}
+
+/// a parameter can be learned t one cc at a time
+/// for a given parameter, find and return the cc that has been learned (if any)
+/// it does this by finding the grid shortcut that corresponds to that param
+/// and then returns what cc or no cc (255) has been mapped to that param shortcut
+int32_t MidiFollow::getCCFromParam(Param::Kind paramKind, int32_t paramID) {
+	for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+		for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
+			bool foundParamShortcut =
+			    (((paramKind == Param::Kind::PATCHED) && (patchedParamShortcuts[xDisplay][yDisplay] == paramID))
+			     || ((paramKind == Param::Kind::UNPATCHED_SOUND)
+			         && (unpatchedParamShortcuts[xDisplay][yDisplay] == paramID))
+			     || ((paramKind == Param::Kind::UNPATCHED_GLOBAL)
+			         && (globalEffectableParamShortcuts[xDisplay][yDisplay] == paramID)));
+
+			if (foundParamShortcut) {
+				return paramToCC[xDisplay][yDisplay];
+			}
+		}
+	}
+	return MIDI_CC_NONE;
+}
+
+/// called from playback handler
+/// determines whether a note message received is midi follow relevant
+/// and should be routed to the active context for further processing
+void MidiFollow::noteMessageReceived(MIDIDevice* fromDevice, bool on, int32_t channel, int32_t note, int32_t velocity,
+                                     bool* doingMidiThru, bool shouldRecordNotesNowNow, ModelStack* modelStack) {
+	// midi follow mode
+	if (midiEngine.midiFollow) {
+		Clip* clip = getClipForMidiFollow(true);
+		if (!on && !clip) {
+			// for note off's when you're no longer in an active clip,
+			// see if a note on message was previously sent so you can
+			// direct the note off to the right place
+			clip = clipForLastNoteReceived;
+		}
+
+		// Only send if not muted - but let note-offs through always, for safety
+		if (clip && (!on || currentSong->isOutputActiveInArrangement(clip->output))) {
```

Comment:
I think it makes more sense to invert this logic - check match, if there is a match on kit or synth channel then get the clip and send it through. That will be faster to return in the common case that it's not a match anyway

Reply:
Done